difftest/dpi_t1rocketemu/src/interconnect.rs

```diff
@@ -0,0 +1,221 @@
+impl RegularMemory {
+  pub fn with_size(size: u32) -> Self {
+    RegularMemory { data: vec![0; size as usize] }
+  }
+}
+
+impl Device for RegularMemory {
+  fn mem_read(&mut self, addr: AddrInfo, data: &mut [u8]) {
+    let mem_data = &self.data[addr.as_range()];
+    data.copy_from_slice(mem_data);
+  }
+
+  fn mem_write(&mut self, addr: AddrInfo, data: &[u8]) {
+    let mem_data = &mut self.data[addr.as_range()];
+    mem_data.copy_from_slice(data);
+  }
+
+  fn mem_write_masked(&mut self, addr: AddrInfo, data: &[u8], mask: &[bool]) {
+    let mem_data = &mut self.data[addr.as_range()];
+    memcpy_mask(mem_data, data, mask);
+  }
+}
+
+fn memcpy_mask(dst: &mut [u8], src: &[u8], mask: &[bool]) {
+  for i in 0..mask.len() {
+    if mask[i] {
+      dst[i] = src[i];
+    }
+  }
+}
+
+pub struct DeviceEntry {
+  base_and_size: (u32, u32),
+  device: Box<dyn Device>,
+}
+
+pub struct AddressSpace {
+  devices: Vec<DeviceEntry>,
+}
+
+impl AddressSpace {
+  pub fn read_mem(&mut self, addr: u32, len: u32, data: &mut [u8]) {
+    assert_eq!(len as usize, data.len());
+    let Some(device_idx) = self.find_device_idx(addr, len) else {
+      panic!("read error (no device found), addr=0x{addr:08x}, len={len}B");
+    };
+
+    let dev_entry = &mut self.devices[device_idx];
+    let addr = AddrInfo { offset: addr - dev_entry.base_and_size.0, len };
+
+    dev_entry.device.mem_read(addr, data);
+  }
+
+  pub fn write_mem(&mut self, addr: u32, len: u32, data: &[u8]) {
+    assert_eq!(len as usize, data.len());
+    let Some(device_idx) = self.find_device_idx(addr, len) else {
+      panic!("write error (no device found), addr=0x{addr:08x}, len={len}B");
+    };
+
+    let dev_entry = &mut self.devices[device_idx];
+    let addr = AddrInfo { offset: addr - dev_entry.base_and_size.0, len };
+
+    dev_entry.device.mem_write(addr, data);
+  }
+
+  pub fn write_mem_masked(&mut self, addr: u32, len: u32, data: &[u8], mask: &[bool]) {
+    assert_eq!(len as usize, data.len());
+    assert_eq!(len as usize, mask.len());
+    let Some(device_idx) = self.find_device_idx(addr, len) else {
+      panic!("write error (no device found), addr=0x{addr:08x}, len={len}B");
+    };
+
+    let dev_entry = &mut self.devices[device_idx];
+    let addr = AddrInfo { offset: addr - dev_entry.base_and_size.0, len };
+
+    dev_entry.device.mem_write_masked(addr, data, mask);
+  }
+
+  fn find_device_idx(&self, addr: u32, len: u32) -> Option<usize> {
+    for (idx, dev) in self.devices.iter().enumerate() {
+      let (base, size) = dev.base_and_size;
+      if base <= addr && addr - base < size {
+        return if addr - base + len < size {
+          Some(idx)
+        } else {
+          None
+        };
+      }
+    }
+
+    None
+  }
+}
+
+/// Memory map:
+/// - 0x0400_0000 - 0x0600_0000 : framebuffer
+/// - 0x1000_0000 - 0x1000_1000 : simctrl
+/// - 0x2000_0000 - 0xc000_0000 : ddr
+/// - 0xc000_0000 - 0xc040_0000 : sram
+/// TODO: simctrl is inside ddr, move it elsewhere
```

Comment:
Fixed?

Reply:
Yes, it's fixed. The inaccurate comment is removed in new commit.